tck/src/main/java/com/akkaserverless/javasdk/tck/model/replicatedentity/TckModelReplicatedEntity.java

```diff
@@ -199,6 +201,51 @@ private void applyUpdate(ReplicatedData replicatedData, Update update) {
             break;
         }
         break;
+      case REPLICATED_COUNTER_MAP:
+        @SuppressWarnings("unchecked")
+        ReplicatedCounterMap<String> counterMap = (ReplicatedCounterMap<String>) replicatedData;
+        switch (update.getReplicatedCounterMap().getActionCase()) {
+          case ADD:
+            String addKey = update.getReplicatedCounterMap().getAdd();
+            counterMap.increment(addKey, 0);
+            break;
+          case UPDATE:
+            String updateKey = update.getReplicatedCounterMap().getUpdate().getKey();
+            long change = update.getReplicatedCounterMap().getUpdate().getChange();
+            counterMap.increment(updateKey, change);
+            break;
+          case REMOVE:
+            String removeKey = update.getReplicatedCounterMap().getRemove();
+            counterMap.remove(removeKey);
+            break;
+          case CLEAR:
+            counterMap.clear();
+            break;
+        }
+        break;
+      case REPLICATED_REGISTER_MAP:
+        @SuppressWarnings("unchecked")
+        ReplicatedRegisterMap<String, String> registerMap =
+            (ReplicatedRegisterMap<String, String>) replicatedData;
```

Comment:
I may be missing something, but why both `ReplicatedCounterMap` and `ReplicatedRegisterMap` have type parameters? 

The proto definitions define the `(K, V)` to be `(String, Long)` and `(String, String)` respectively.

True that I still need to wrap my mind around the whole replicate entity feature and get a better view on how that will be used, but at first sight my understanding is that we are already bound to `(String, Long)` and `(String, String)`

Reply:
The key can be anything serializable — protobuf message, json, or primitive. Same for register values. The proto definitions here are just for the TCK tests, which are simplified to only string keys and string values for registers.

Reply:
I understood this was the TCK, but the map entry in the proto is defined as 

```proto
message ReplicatedRegisterMapEntryUpdate {
  string key = 1;
  string value = 2;
}
```

How could we send anything other then String? Are we sending the bytes encoded as base64?

Reply:
That's the protocol for the TCK — which sends messages on how the entities should be updated, and only uses strings.

https://github.com/lightbend/akkaserverless-framework/blob/master/protocols/tck/src/main/protobuf/akkaserverless/tck/model/replicatedentity/replicated_entity.proto

The actual proxy protocol for replicated maps uses protobuf Any:

```proto
message ReplicatedMapEntryDelta {
  // The entry key.
  google.protobuf.Any key = 1;
  ReplicatedEntityDelta delta = 2;
}

message ReplicatedRegisterDelta {
  google.protobuf.Any value = 1;
  ReplicatedEntityClock clock = 2;
  int64 custom_clock_value = 3;
}
```

https://github.com/lightbend/akkaserverless-framework/blob/613d873529f8e0785218b0357c281a30a99c9d31/protocols/proxy/src/main/protobuf/akkaserverless/component/replicatedentity/replicated_entity.proto#L132-L136

https://github.com/lightbend/akkaserverless-framework/blob/613d873529f8e0785218b0357c281a30a99c9d31/protocols/proxy/src/main/protobuf/akkaserverless/component/replicatedentity/replicated_entity.proto#L119-L123

Reply:
Oh, that explains everything. Thanks for clarifying.